src/Utility/ChatTools.cs

```diff
@@ -0,0 +1,296 @@
+            completionOptions.Tools.Add(tool);
+        return completionOptions;
+    }
+
+    /// <summary>
+    /// Converts the tools collection to <see cref="ChatCompletionOptions"/>, filtered by relevance to the given prompt.
+    /// </summary>
+    /// <param name="prompt">The prompt to find relevant tools for.</param>
+    /// <param name="maxTools">The maximum number of tools to return. Default is 3.</param>
+    /// <param name="minVectorDistance">The similarity threshold for including tools. Default is 0.29.</param>
+    /// <returns>A new <see cref="ChatCompletionOptions"/> containing the most relevant tools.</returns>
+    public async Task<ChatCompletionOptions> CreateCompletionOptionsAsync(string prompt, int maxTools = 3, float minVectorDistance = 0.29f)
+    {
+        if (!CanFilterTools)
+            return CreateCompletionOptions();
+
+        var completionOptions = new ChatCompletionOptions();
+        foreach (var tool in await FindRelatedTools(true, prompt, maxTools, minVectorDistance).ConfigureAwait(false))
+            completionOptions.Tools.Add(tool);
+        return completionOptions;
+    }
+
+    private async Task<IEnumerable<ChatTool>> FindRelatedTools(bool async, string prompt, int maxTools, float minVectorDistance)
+    {
+        if (!CanFilterTools)
+            return _tools;
+
+        return (await FindVectorMatches(async, prompt, maxTools, minVectorDistance).ConfigureAwait(false))
+            .Select(e => ParseToolDefinition(e.Data));
+    }
+
+    private async Task<IEnumerable<VectorbaseEntry>> FindVectorMatches(bool async, string prompt, int maxTools, float minVectorDistance)
+    {
+        var vector = async ?
+            await ToolsUtility.GetEmbeddingAsync(_client, prompt).ConfigureAwait(false) :
+            ToolsUtility.GetEmbedding(_client, prompt);
+
+        lock (_entries)
+        {
+            return ToolsUtility.GetClosestEntries(_entries, maxTools, minVectorDistance, vector);
+        }
+    }
+
+    /// <summary>
+    /// Implicitly converts ChatTools to <see cref="ChatCompletionOptions"/>.
+    /// </summary>
+    /// <param name="tools">The ChatTools instance to convert.</param>
+    public static implicit operator ChatCompletionOptions(ChatTools tools) => tools.CreateCompletionOptions();
+
+    internal string CallLocal(ChatToolCall call)
+    {
+        var arguments = new List<object>();
+        if (call.FunctionArguments != null)
+        {
+            ToolsUtility.ParseFunctionCallArgs(call.FunctionArguments, out arguments);
+        }
+        return CallLocal(call.FunctionName, [.. arguments]);
+    }
+
+    private string CallLocal(string name, object[] arguments)
+    {
+        if (!_methods.TryGetValue(name, out MethodInfo method))
+            return $"I don't have a tool called {name}";
+
+        object result = method.Invoke(null, arguments);
+        return result?.ToString() ?? string.Empty;

```

Comment:
Instead of returning string.Empty, this should probably throw an exception explaining that the result of a function cannot be null and the user needs to fix their function.

Reply:
Also, dont we support tools that do things without returning anything?

Reply:
Yes, all tools must return something. If a function has no "useful" return value, the recommendation is to return a string to indicate success or failure (e.g., "success" or "failure"). Ultimately, though, I think the specifics should be up to the user (so that they can tailor the implementation for their specific scenario) instead of us making a decision for them. This is why I think throwing an exception here instead of silently returning `string.Empty` is the correct thing to do.